### PR TITLE
8269039: Disable SHA-1 Signed JARs

### PR DESCRIPTION
**Pull request by GoeLin**
src/java.base/share/conf/security/java.security
Does not resolve because 11 mentions "include jdk.disabled.namedCurves"

src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java
Some hunks did not apply because DISABLED_CHECK was renamed
to JAR_DISABLED_CHECK in 17.
Other hunks patch methods not in 11: checkWeakKey(), checkWeakAlg()
as well as the calls to these methods.

test/jdk/java/security/Security/signedfirst/Dyn.sh
test/jdk/java/security/Security/signedfirst/Static.sh
Deleting did not apply.

test/jdk/java/util/jar/JarInputStream/signed.jar
Patching this binary file failed. I just copied
the file from 17.

test/jdk/sun/security/tools/jarsigner/CheckSignerCertChain.java
Patch skipped, test not in 11.

test/jdk/sun/security/tools/jarsigner/TimestampCheck.java
Resolved. Checked output differed.

test/lib/jdk/test/lib/security/SecurityUtils.java
The change to this file was already backported.

In addition, I adapted 
sun/security/tools/jarsigner/DefaultOptions.java
sun/security/tools/jarsigner/NameClash.java
sun/security/tools/jarsigner/EC.java
according to 
 "8172404: Tools should warn if weak algorithms are used before restricting them"
which makes the tests pass.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8269039](https://bugs.openjdk.org/browse/JDK-8269039): Disable SHA-1 Signed JARs
 * [JDK-8264362](https://bugs.openjdk.org/browse/JDK-8264362): Disable SHA-1 Signed JARs (**CSR**)

### Reviewers
 * [Martin Balao](https://openjdk.org/census#mbalao) (@martinuy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev pull/1244/head:pull/1244` \
`$ git checkout pull/1244`

Update a local copy of the PR: \
`$ git checkout pull/1244` \
`$ git pull https://git.openjdk.org/jdk11u-dev pull/1244/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1244`

View PR using the GUI difftool: \
`$ git pr show -t 1244`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/1244.diff">https://git.openjdk.org/jdk11u-dev/pull/1244.diff</a>

</details>

### CODE REVIEW

**Comment by GoeLin** [comment]
@mbalao @martinuy 
Hi, could you have a look at this change? I would appreciate a review by someone with experience with this code. Thanks!

**Review by martinuy — APPROVED** [review]
Hi Goetz,

Thanks for proposing this backport.

Comments:

 * src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java
   * Looks to me that the parameter 'CertPathConstraintsParameters cpcp' is added to the method printCert but not used. In 17u, this parameter is passed to checkWeakKey (introduced in 8259401), which has not been backported to 11u. I'd consider either removing it or backporting 8259401 too, which does not seem (at first glance) too complex. Give that this is a minor comment and that we are targeting 11.0.17, this suggestion won't be a blocker.

 * test/jdk/java/security/Security/signedfirst/exp.jar
   * I was comparing against 17u-dev and found that this file was not deleted there. The correct action is to delete it, as proposed in the 11u backport and in the original patch. When the test was a script, this file was used. Now that the test is in Java, it uses the test libs to create the JAR in run time. No harm but, technically, we should delete this file from 17u.
     * Note: this comment also applies to the file test/jdk/java/security/Security/signedfirst/keystore.jks.

In the test/jdk/sun/security/tools/jarsigner directory, I've not found references to "SHA1" that are not in 17u. The changes for tests removed in later releases look good to me.

I've also checked that there are no regressions in the following test categories:

 * test/jdk/java/security/Security/signedfirst
 * test/jdk/sun/security/tools/jarsigner

In summary, the 11u backport looks good to me. Good job.

I've seen that the 17u backport links the original CSR. In the old days we used to have a CSR specific for the backport. Please check that with the 11u release maintainers.

Thanks,
Martin.-

--
[1] - https://github.com/openjdk/jdk11u-dev/pull/1244/commits/942e3c297b1e99799d1b8443a4871bc3c1997741#diff-7083af3b8473a092987afa0bbb4d1694664649534bac716f6b9cd3c3b9833219R1449

**Comment by GoeLin** [comment]
Hi Martin,
thanks for having a look that fast. 
* Yes, I added the wrong CSR. Fixed for 17.
* I will remove the file in 17.
* I'll have a look at 8259401 and backport if feasible.

**Comment by GoeLin** [comment]
Hi Martin,
I filed 8291595: [17u] Delete files missed in backport of 8269039 
Also, I had a look at 8259401. It can be backported clean and passes our testing. But it comes with a release note about the warning, and there is no such note for 11. So I think we should rather not backport this change. So should I remove the argument?
On the other side, eventual follow up backports apply better if I keep it as-is, and this should not harm performance as it is only called for printing.

**Comment by GoeLin** [comment]
Having double-checked with my colleagues I would like to push it as-is.
